### PR TITLE
Reduce duplication in AnyInstanceMethod class

### PR DESCRIPTION
**Pull request by nitishr**
As suggested in #359.

Also addresses #332

Build is currently failing due to minitest latest version not supporting ruby 1.8.6 as described [here](https://github.com/freerange/mocha/pull/370#issuecomment-537675002)

### CODE REVIEW

**Comment by floehopper** [comment]
@nitishr I've now fixed the minitest compatibility issue in #373. Can you rebase this branch against `master` so we can see whether all the tests pass. Many thanks for your work on this.

File: lib/mocha/any_instance_method.rb [hunk 1]
```diff
@@ -13,8 +13,10 @@ def reset_mocha
 
     def restore_original_method
       return if use_prepended_module_for_stub_method?
-      return unless stub_method_overwrites_original_method?
-      original_method_owner.send(:define_method, method_name, original_method)
+      if stub_method_overwrites_original_method?
+        original_method_owner.send(:define_method, method_name, original_method)
+      end
+      return unless original_visibility
```
Comment (floehopper) [inline lib/mocha/any_instance_method.rb]: Can you explain whether these changes are a pure refactoring or do they potentially change the behaviour even if they don't break any tests? Either way it would be helpful to expand the commit message to explain a bit more about how you are able to make this change with confidence!

File: Gemfile [hunk 2]
```diff
@@ -1,3 +1,5 @@
 source 'https://rubygems.org'
 
 gemspec
+
+gem 'pry', '0.9.12.6'
```
Comment (floehopper) [inline Gemfile]: I assume you left this in by accident. Please can you remove it?

File: lib/mocha/any_instance_method.rb [hunk 3]
```diff
@@ -11,17 +11,12 @@ def reset_mocha
...
-      return unless original_visibility
-      Module.instance_method(original_visibility).bind(original_method_owner).call(method_name)
-    end
-
     private
 
+    def original_method_body
+      original_method
+    end
+
```
Comment (floehopper) [inline lib/mocha/any_instance_method.rb]: Am I right in thinking this has potentially _fixed_ some un-tested behaviour in Ruby < v1.9? If so, can you explain this in the commit note?

File: lib/mocha/class_method.rb [hunk 4]
```diff
@@ -102,6 +95,11 @@ def method_visibility
 
     attr_reader :original_method, :original_visibility
 
+    def original_method_body
+      original_method_in_scope = original_method
+      PRE_RUBY_V19 ? proc { |*args, &block| original_method_in_scope.call(*args, &block) } : original_method
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Minor: I'm not super-keen on using the ternary operator here, because the line ends up being quite long.

File: lib/mocha/class_method.rb [hunk 5]
```diff
@@ -125,13 +125,8 @@ def use_prepended_module_for_stub_method
...
-    def stub_method_definition
-      method_implementation = <<-CODE
-      def #{method_name}(*args, &block)
-        mocha.method_missing(:#{method_name}, *args, &block)
-      end
-      CODE
-      [method_implementation, __FILE__, __LINE__ - 4]
+    def stub_method_body(method_name)
+      proc { |*args, &block| mocha.method_missing(method_name, *args, &block) }
     end
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Nice! 👏

File: lib/mocha/class_method.rb [hunk 6]
```diff
@@ -108,7 +106,7 @@ def store_original_method_visibility
       @original_visibility = method_visibility
     end
 
-    def stub_method_overwrites_original_method?
+    def stub_method_overwrites_original?
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: I'd prefer if the renaming of this method could be moved into a separate commit.

File: lib/mocha/class_method.rb [hunk 7]
```diff
@@ -66,11 +63,8 @@ def remove_new_method
 
     def restore_original_method
       return if use_prepended_module_for_stub_method?
-      if stub_method_overwrites_original_method?
-        original_method_owner.send(:define_method, method_name, original_method_body)
-      end
-      return unless original_visibility
-      Module.instance_method(original_visibility).bind(original_method_owner).call(method_name)
+      original_method_owner.send(:define_method, method_name, original_method_body) if stub_method_overwrites_original?
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Again, I'm not very keen on these long lines, especially when the condition is on the end of the line, because I think it's easy to miss. I should really setup a Rubocop rule to enforce this!

Also it would be nice to keep stylistic changes like this separate from the core refactoring here.

File: lib/mocha/class_method.rb [hunk 8]
```diff
@@ -93,6 +87,10 @@ def method_visibility
 
     private
 
+    def maintain_original_visibility(method_owner)
+      Module.instance_method(original_visibility).bind(method_owner).call(method_name) if original_visibility
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Another long line with a condition on the end.

File: lib/mocha/any_instance_method.rb [hunk 9]
```diff
@@ -28,5 +20,9 @@ def stub_method_body(method_name)
     def original_method_owner
       stubbee
     end
+
+    def instance
```
Comment (floehopper) [inline lib/mocha/any_instance_method.rb]: I'm not sure about the method name, `#instance`, either. Maybe something like `#mock_owner` would be better...? What do you think?

File: lib/mocha/any_instance_method.rb [hunk 10]
```diff
@@ -3,6 +3,10 @@
 
 module Mocha
   class AnyInstanceMethod < ClassMethod
+    def instance
+      stubbee.any_instance
+    end
+
```
Comment (floehopper) [inline lib/mocha/any_instance_method.rb]: Why did you need to make this method public?

File: lib/mocha/class_method.rb [hunk 11]
```diff
@@ -85,6 +88,10 @@ def method_visibility
     end
     alias_method :method_defined_in_stubbee_or_in_ancestor_chain?, :method_visibility
 
+    def instance
+      stubbee
+    end
+
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Why did you need to make this method public?

File: lib/mocha/class_method.rb [hunk 12]
```diff
@@ -53,7 +53,10 @@ def hide_original_method
     end
 
     def define_new_method
-      stub_method_owner.send(:define_method, method_name, stub_method_body(method_name))
+      self_in_scope = self
+      stub_method_owner.send(:define_method, method_name) do |*args, &block|
+        self_in_scope.instance.mocha.method_missing(self_in_scope.method_name, *args, &block)
+      end
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: It looks as if this means you can remove the implementation of `ClassMethod#stub_method_body`, right?

File: lib/mocha/class_method.rb [hunk 13]
```diff
@@ -53,7 +53,10 @@ def hide_original_method
     end
 
     def define_new_method
-      stub_method_owner.send(:define_method, method_name, stub_method_body(method_name))
+      self_in_scope = self
+      stub_method_owner.send(:define_method, method_name) do |*args, &block|
+        self_in_scope.instance.mocha.method_missing(self_in_scope.method_name, *args, &block)
```
Comment (floehopper) [inline lib/mocha/class_method.rb]: Nice! 👏

**Comment by nitishr** [comment]
@floehopper ready for a review again

File: test/acceptance/stub_any_instance_method_test.rb [hunk 14]
```diff
@@ -43,6 +43,28 @@ def self.public(*args); end
...
+    assert_snapshot_unchanged(instance) do
+      test_result = run_as_test do
+        klass.any_instance.stubs(:my_included_method).returns(:new_return_value)
+        assert_equal :new_return_value, instance.my_included_method
+      end
+      assert_passed(test_result)
+    end
+    assert_equal :original_return_value, instance.my_included_method
+  end
+
```
Comment (floehopper) [inline test/acceptance/stub_any_instance_method_test.rb]: Thanks so much for adding this test! 👍

**Comment by floehopper** [comment]
@nitishr 

Thanks for making the improvements I suggested. I hope you don't mind, but I've curated your commits onto [a new branch](https://github.com/freerange/mocha/compare/reduce-duplication-in-any-instance-method-class). The main changes are that I've split up some of the early commits a bit more for clarity and I've expanded many of the commit notes to try to make things clearer and more consistent. I found this was a useful way for me to review the changes and get confidence that we're not breaking anything by making them.

I added you as a co-author to the new commits I added and retained your authorship on the commits I amended. If it's OK with you, I'll create a new PR from the new branch and get that merged. However, I want to make sure you're happy that I've fully credited your work and so let me know if you'd prefer I took a different approach.

Thanks again for taking on this work. It's something I've been meaning to do for ages and I've found it really motivating to have you work on it!

**Comment by nitishr** [comment]
This is great, @floehopper. I don't mind at all, and I appreciate that your commits are a lot cleaner and clearer.

It looks like my build is now broken/stuck for ruby 2.0 & 2.1 on travis. I'm currently investigating that.

**Comment by nitishr** [comment]
It's the new test that's hanging the build.

**Comment by floehopper** [comment]
> It's the new test that's hanging the build.

Yes, I just saw that. Running it locally, it seems as if it hangs even in the commit where the new test was added, which at least rules out the subsequent commits. Puzzling! 😕

**Comment by floehopper** [comment]
The Travis build is failing for the curated branch too: https://travis-ci.org/freerange/mocha/builds/594232934.

**Comment by floehopper** [comment]
Superseded by #378. Closing.
